relay/messenger.go

```diff
@@ -0,0 +1,136 @@
+package relay
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/vultisig/mobile-tss-lib/tss"
+)
+
+type MessengerImp struct {
+	Server    string
+	SessionID string
+}
+
+func (m *MessengerImp) Send(from, to, body string) error {
+	hash := md5.New()
+	hash.Write([]byte(body))
+	hashStr := hex.EncodeToString(hash.Sum(nil))
+
+	if hashStr == "" {
+		return fmt.Errorf("hash is empty")
+	}
+
+	buf, err := json.MarshalIndent(struct {
+		SessionID string   `json:"session_id,omitempty"`
+		From      string   `json:"from,omitempty"`
+		To        []string `json:"to,omitempty"`
+		Body      string   `json:"body,omitempty"`
+		Hash      string   `json:"hash,omitempty"`
+	}{
+		SessionID: m.SessionID,
+		From:      from,
+		To:        []string{to},
+		Body:      body,
+		Hash:      hashStr,
+	}, "", "  ")
+	if err != nil {
+		return fmt.Errorf("fail to marshal message: %w", err)
+	}
+
+	url := fmt.Sprintf("%s/message/%s", m.Server, m.SessionID)
+	req, err := http.NewRequest("POST", url, bytes.NewReader(buf))
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	if body == "" {
+		return fmt.Errorf("body is empty")
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.Status != "202 Accepted" {
+		return fmt.Errorf("fail to send message, response code is not 202 Accepted: %s", resp.Status)
+	}
+
+	fmt.Println("message sent")
+
+	return nil
+}
+
+func DownloadMessage(server, session, key string, tssServerImp tss.Service, endCh chan struct{}, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case <-endCh: // we are done
+			return
+		case <-time.After(time.Second):
+			resp, err := http.Get(server + "/message/" + session + "/" + key)
+			if err != nil {
+				log.Println("fail to get data from server:", err)
+				continue
+			}
+			if resp.StatusCode != http.StatusOK {
+				log.Println("fail to get data from server:", resp.Status)
+				continue
+			}
+			decoder := json.NewDecoder(resp.Body)
+			var messages []struct {
+				SessionID string   `json:"session_id,omitempty"`
+				From      string   `json:"from,omitempty"`
+				To        []string `json:"to,omitempty"`
+				Body      string   `json:"body,omitempty"`
+			}
+			if err := decoder.Decode(&messages); err != nil {
+				if err != io.EOF {
+					log.Println("fail to decode messages:", err)
+				}
+				continue
+			}
+			for _, message := range messages {
+				if message.From == key {
+					continue
+				}
+
+				hash := md5.Sum([]byte(message.Body))
```

Comment:
You also need to implement a in memory cache to keep track the message that has been applied before , see the implementation in IOS or android https://github.com/vultisig/vultisig-android/blob/main/app/src/main/java/com/vultisig/wallet/tss/TssMessagePuller.kt#L62